cassandra-analytics-integration-framework/src/main/java/org/apache/cassandra/sidecar/testing/SharedClusterIntegrationTestBase.java

```diff
@@ -0,0 +1,417 @@
+ *     <li>(Optional) Before cluster shutdowns (implementer can supply)
+ *     <li>Close cluster
+ *     <li>(Optional) Before tear down ends (implementer can supply)
+ * </ol>
+ */
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+@ExtendWith(VertxExtension.class)
+public abstract class SharedClusterIntegrationTestBase
+{
+    protected final Logger logger = LoggerFactory.getLogger(SharedClusterIntegrationTestBase.class);
+
+    protected Vertx vertx;
+    protected DnsResolver dnsResolver;
+    protected AbstractCluster<? extends IInstance> cluster;
+    protected Server server;
+    protected Injector injector;
+
+    static
+    {
+        // Initialize defaults to configure the in-jvm dtest
+        TestUtils.configureDefaultDTestJarProperties();
+    }
+
+    @BeforeAll
+    protected void setup() throws InterruptedException, IOException
+    {
+        Optional<TestVersion> testVersion = TestVersionSupplier.testVersions().findFirst();
+        assertThat(testVersion).isPresent();
+        logger.info("Testing with version={}", testVersion);
+        cluster = provisionCluster(testVersion.get());
+        assertThat(cluster).isNotNull();
+        initializeSchemaForTest();
+        startSidecar(cluster);
+        beforeTestStart();
+    }
+
+    @AfterAll
+    protected void tearDown() throws InterruptedException
+    {
+        beforeSidecarStop();
+        stopSidecar();
+        beforeClusterShutdown();
+        closeCluster();
+        beforeTearDownEnd();
+    }
+
+    /**
+     * @param testVersion the Cassandra version to use for the test
+     * @return a provisioned cluster to use for tests
+     * @throws IOException when provisioning a cluster fails
+     */
+    protected abstract UpgradeableCluster provisionCluster(TestVersion testVersion) throws IOException;
+
+    /**
+     * Initialize required schemas for the tests upfront before the test starts
+     */
+    protected abstract void initializeSchemaForTest();
+
+    /**
+     * Override to perform an action before the tests start
+     */
+    protected void beforeTestStart()
+    {
+    }
+
+    /**
+     * Override to perform an action before Sidecar stops
+     */
+    protected void beforeSidecarStop()
+    {
+    }
+
+    /**
+     * Override to perform an action before the cluster stops
+     */
+    protected void beforeClusterShutdown()
+    {
+    }
+
+    /**
+     * Override to perform an action as the last step of the tear down method
+     */
+    protected void beforeTearDownEnd()
+    {
+    }
+
+    protected void createTestKeyspace(QualifiedName name, Map<String, Integer> rf)
+    {
+        createTestKeyspace(name.maybeQuotedKeyspace(), rf);
+    }
+
+    protected void createTestKeyspace(String keyspace, Map<String, Integer> rf)
+    {
+        cluster.schemaChangeIgnoringStoppedInstances("CREATE KEYSPACE IF NOT EXISTS " + keyspace
+                                                     + " WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', " +
```

Comment:
I guess it is same to assume that we always want to test with `NetworkTopologyStrategy`

Reply:
yeah, I didn't change this as all tests are currently using `NetworkTopologyStrategy`, but this is open to modification in the future as needed